### DISCUSSION TITLE
Body embed parameter and documentation

### DISCUSSION THREAD

**Discussion by MaelAbgrall** [fastapi/fastapi#6789] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [ ] I already read and followed all the tutorial in the docs and didn't find an answer.
* [ ] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [ ] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [ ] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
from fastapi import FastAPI
from pydantic import BaseModel, constr

app = FastAPI()

class CategoryBase(BaseModel):
  title: constr(min_length=4, max_length=30, regex="^[a-zA-Z0-9 _-]+$")  # type: ignore
  description: Optional[str]


class CategoryCreate(CategoryBase):
  pass


class CategoryPublic(CategoryBase):
  pass


@app.post("/", response_model=CategoryPublic, name="create-category", status_code=status.HTTP_201_CREATED)
async def create_category(
  new_category: CategoryCreate = Body(..., embed=True),
) -> CategoryPublic:
  """I'm the culprit of weird documentation."""
  # some code ...


@app.post("/", response_model=CategoryPublic, name="create-category", status_code=status.HTTP_201_CREATED)
async def create_category(
  new_category: CategoryCreate = Body(..., embed=False),
) -> CategoryPublic:
  """I'm not."""
  # some code ...
```

### Description

*TLDR:*
How to have a nice documentation with the Body(..., embed=True)

I'm trying to have a nice documentation for the api I'm creating, and I noticed that when using embed=True for the Body parameter, a new schema is created. This not only create two schemas (one for the Pydantic model, and one for the Body), but the name used are not very readable.

I tried to look in the documentation, but I did not see anything about changing the name, or (better) simply preventing the documentation of a somewhat duplicate schema

### Environment

* OS: Docker image of python:3.8

* FastAPI Version: 0.63.0

### Additional context

![fastapi](https://user-images.githubusercontent.com/25052677/115381628-a7c3eb80-a1cb-11eb-8d72-0530a4f59c5d.png)

**Comment by frankie567** [comment, marked as answer]
# Rationale 
The "duplicate" schema named `Body_create_category_api...` is created because you set the option `embed` to `True`. The effect of this option is to "nest" the data inside its own key (see https://fastapi.tiangolo.com/tutorial/body-multiple-params/?h=embed#embed-a-single-body-parameter).

That's why FastAPI creates a new OpenAPI schema. This schema:

```json
{
    "title": "string",
    "description": "string"
}
```

is **different** from the nested one:

```json
{
    "new_category": {
        "title": "string",
        "description": "string"
    }
}
```

I've looked at the FastAPI internals, and it seems that the schema generated for `Body` always use the operation unique id and cannot be customized: https://github.com/tiangolo/fastapi/blob/22528373bba6a654323de416ad5c867cbadb81bb/fastapi/routing.py#L386

# Recommended solution

Now, for your use-case, unless you need to have another object besides the category, I suggest that you simply don't use embed, like this:

```py
@app.post("/", response_model=CategoryPublic, name="create-category", status_code=status.HTTP_201_CREATED)
async def create_category(new_category: CategoryCreate) -> CategoryPublic:
  """I'm not."""
  # some code ...
```

# Prevent the weird name

Now, if you really want to embed your schema and avoid the `Body_create_...` thing, I guess you'll have to write the Pydantic model yourself:

```py
class EmbedCategoryCreate(BaseModel):
    new_category: CategoryCreate


@app.post("/", response_model=CategoryPublic, name="create-category", status_code=status.HTTP_201_CREATED)
async def create_category(embed_new_category: EmbedCategoryCreate) -> CategoryPublic:
  """I'm not."""
  new_category = embed_new_category.new_category
  # some code ...
```

But this may be a bit circumvoluted.

**Comment by MaelAbgrall** [comment]
Not being able to rename is unfortunate... I'm not an experienced software engineer, but I suppose it's not rare to use multiple body objects?

I'll have some refactoring to do in the meantime, thank you!

**Comment by frankie567** [comment]
Well, in a REST API, we usually consider that one endpoint == one resource, so it's *very* common to have something like this:

```
POST /category

{
    "title": "Foo",
    "description": "Bar"
}
```

Getting or posting several objects in one request is much rarer, and generally not consisered as "RESTful".

**Comment by tiangolo** [comment]
Thanks for the help here @frankie567 ! 👏  🙇 

Thanks for reporting back and closing the issue @MaelAbgrall  👍

Yep, customizing the autogenerated name is still not supported, but maybe in a future version.

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
